[16.01] Allow admin to serialize/deserialize dataset permissions

carlfeberhard opened the pull request:
Upshot: admin can now (16.01) properly get and set permissions over the API.
- Adds `unit/manager/test_DatasetManager` tests for serialization and deserialization by admin users. 
- Updates `managers/datasets.py` to add `is_admin` check to `serialize_permissions`
- (deserialization for admin has been possible (`error_unless_permitted`), but unwise if you can't get the serialized form in the first place.

**getting**:
    // GET https://galaxy/api/histories/:history_id/contents/:hda_id?keys=id,permissions

```
{
  "id": "0778472e8df1f331",
  "permissions": {
    //public
    "access": [

    ],
    //role id of private role for me
    "manage": [
      "54d6187b0652a2a1"
    ]
  }
}
```

**setting**: POST (add/subtract role ids from the `access` and `manage` arrays above.)
    // POST https://galaxy/api/histories/:history_id/contents/:hda_id?keys=id,permissions

```
// post body
{
  "permissions": {
    //will now be private to me
    "access": [
      "54d6187b0652a2a1"
    ],
    //role id of private role for me
    "manage": [
      "54d6187b0652a2a1"
    ]
  }
}
```

carlfeberhard committed:
Datasets: allow admin to serialize permissions

And updates tests.

carlfeberhard committed:
Datasets: confirm admin ability to set permissions

The power was inside you all along.

martenson commented on test/unit/managers/test_DatasetManager.py:
```diff
@@ -162,6 +162,11 @@ def test_create_private_dataset( self ):
         self.log( "a private dataset shouldn't be accessible to just anyone" )
         self.assertFalse( self.dataset_manager.permissions.access.is_permitted( dataset, user3 ) )
 
+        self.log( "a private dataset shouldn be manageable by an admin" )
```
s/shouldn/shouldn't

martenson commented:
👍

carlfeberhard committed:
Permissions, datasets: allow admin in RBAC

- Admin are considered permitted when calling
  ** dataset_manager.permssions.access and
  ** dataset_manager.permissions.manage
- Update bad test
- Add admin test to other permission checks

carlfeberhard committed:
Datasets: test for anon access/manage

carlfeberhard committed:
Datasets: test for anon deserialization

carlfeberhard commented:
Updated to allow admin permission from within the `rbac_secured.py` classes themselves. Updated tests.

martenson commented:
I am still 👍 here

mvdbeek commented on lib/galaxy/managers/rbac_secured.py:
```diff
@@ -173,6 +173,9 @@ def is_permitted( self, dataset, user ):
         # anonymous users cannot manage permissions on datasets
         if self.user_manager.is_anonymous( user ):
             return False
+        # admin can always manager permissions
```
manager is a typo, right?

mvdbeek commented on lib/galaxy/managers/rbac_secured.py:
```diff
@@ -225,7 +228,9 @@ def is_permitted( self, dataset, user ):
         #   anonymous access to public datasets
-        return ( self._is_public_from_roles( current_roles ) or
+        return ( self._is_public_based_on_roles( current_roles ) or
+                 # admin can always manager permissions
```
same as above

mvdbeek commented:
👍

carlfeberhard committed:
Dataset perms: correct comment

carlfeberhard commented:
Updated. @martenson @mvdbeek Thanks for the review
